src/Interpreters/executeQuery.cpp

```diff
@@ -626,6 +631,27 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
         }
         else
         {
+            /// We need to start the (implicit) transaction before getting the interpreter as this will get links to the latest snapshots
+            if (!context->getCurrentTransaction() && settings.implicit_transaction && !ast->as<ASTTransactionControl>())
+            {
+                try
+                {
+                    /// If there is no session (which is the default for the HTTP Handler), set up one just for this as it is necessary
+                    /// to control the transaction lifetime
+                    if (!context->hasSessionContext())
+                        context->makeSessionContext();
```

Comment:
Let's also add an assertion (or even throw `LOGICAL_ERROR`) if current context is not a query context either. It will protect from creating a transaction in global context (when we call `executeQuery` internally).